[Zhu Shiqi] iP

SQ77 opened the pull request:
# Oliver
> “If we want users to like our software, we should design it to behave like a likeable person.” -Alan Cooper ([source](https://kartaca.com/en/inspiring-quotes-for-software-developers/))

Oliver frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use 😮 

All you need to do is,
1. Download the jar file from [here](https://github.com/SQ77/ip/releases/tag/v0.3)
2. Open your terminal
3. Run the command `java -jar "{filename}.jar"`
4. Let it manage your tasks for you 🪄 🌟 

And it is **FREE!**

Features:
- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SQ77 committed:
skeletal version

SQ77 committed:
Echo commands by user

SQ77 committed:
Add, List

SQ77 committed:
Task class

SQ77 committed:
Add the ability to mark tasks as done

SQ77 committed:
Add three types of tasks

SQ77 committed:
ToDos, Events, Deadlines

SQ77 committed:
Fix bug in todo adding

SQ77 committed:
Semi-automate the testing of Oliver

SQ77 committed:
Add error handling

SQ77 committed:
Add support for deleting tasks

SQ77 committed:
Refactor code to make it more readable

SQ77 committed:
Add method for formatting tasks in data file

SQ77 committed:
Add support for saving tasks to hard disk

SQ77 committed:
Add support for saving tasks to hard disk

SQ77 committed:
Change storage of date and times

SQ77 committed:
Teach oliver to understand dates and times

SQ77 committed:
Merge branch 'branch-Level-8'

SQ77 committed:
Add new classes to extract out closely related code

SQ77 committed:
Refactor code to include usage of new classes

SQ77 committed:
Merge branch 'A-MoreOOP'

SQ77 committed:
Put all classes into the oliver package

SQ77 committed:
Merge branch 'A-Packages'

SQ77 committed:
Merge branch 'add-gradle-support'

SQ77 committed:
Add JUnit tests for Parser and TaskList

SQ77 committed:
Add support for creating empty taskList

SQ77 committed:
Modify build.gradle by changing the defaults

SQ77 committed:
Change helper methods to private

SQ77 committed:
Add support for searching for tasks with keyword

SQ77 committed:
Merge branch 'branch-Level-9'

SQ77 committed:
Tweak the code to comply with coding standards

SQ77 committed:
Merge branch 'branch-A-CodingStandard'

SQ77 committed:
Add JavaDoc comments to the code

SQ77 committed:
Merge branch 'branch-A-JavaDoc'

SQ77 committed:
Change print statements to return strings

SQ77 committed:
Add basic javaFX GUI for oliver

SQ77 committed:
Change default images of oliver and user

SQ77 committed:
Add css styling to the GUI

SQ77 committed:
Merge branch 'branch-Level-10'

SQ77 committed:
Make minor changes to output format

SQ77 committed:
Improve readability of code

appleshill commented on src/main/java/oliver/MainWindow.java:
```diff
@@ -0,0 +1,67 @@
+        userInput.clear();
+
+        if (input.equalsIgnoreCase("bye")) {
+            // Exit the application after 2 seconds of delay
```
I like the use of delay

appleshill commented on src/main/java/oliver/Oliver.java:
```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    public void run() {
```
You may want to include some JavaDocs for this and getResponse method

appleshill commented on src/main/java/oliver/Oliver.java:
```diff
@@ -0,0 +1,205 @@
+        } catch (IndexOutOfBoundsException e) {
+            return ui.showMissingArgsError();
+        } catch (DateTimeParseException e) {
+            return "Invalid date or time. Please enter the date and time in the following format: YYYY-MM-DD HHmm";
```
Beyond the soft limit of line length, you may want to consider adding a line break.

appleshill commented on src/main/java/oliver/Storage.java:
```diff
@@ -0,0 +1,91 @@
+            }
+            writer.close();
+        } catch (IOException e) {
+            System.out.println("\tError occurred when writing to the data file.");
```
Maybe you could enhance this with the filePath to give a better indication of what's going on when the error occurs?

appleshill reviewed (commented):
Generally well done, only a few very minor issues and improvements

jeanyang04 commented on src/main/java/oliver/Oliver.java:
```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    public String getResponse(String input) {
+        String command = Parser.parseCommand(input);
+        if (input.equalsIgnoreCase("bye")) {
+            storage.writeToFile(this.tasks);
+            ui.close();
+            this.isRunning = false;
+            return ui.showBye();
+        } else if (input.equalsIgnoreCase("list")) {
+            return ui.showList(this.tasks);
+        } else if (command.equalsIgnoreCase("mark")) {
+            return handleMark(input);
+        } else if (command.equalsIgnoreCase("unmark")) {
+            return handleUnmark(input);
+        } else if (command.equalsIgnoreCase("todo")) {
+            return handleTodo(input);
+        } else if (command.equalsIgnoreCase("deadline")) {
+            return handleDeadline(input);
+        } else if (command.equalsIgnoreCase("event")) {
+            return handleEvent(input);
+        } else if (command.equalsIgnoreCase("delete")) {
+            return handleDelete(input);
+        } else if (command.equalsIgnoreCase("find")) {
+            return ui.showSearch(this.tasks, Parser.parseArgs(input));
+        } else {
+            return "Invalid command. Command was not recognised.";
+        }
+    }
```
Can a switch statement or enum be used to handle these if-else statements to make your code cleaner?

jeanyang04 commented on src/main/java/oliver/Ui.java:
```diff
@@ -0,0 +1,165 @@
+     * @param numTasks the number of tasks in the list
+     * @return string representation of the confirmation message
+     */
+    public String showAdd(Task addedTask, int numTasks) {
```
Could this method name be more descriptive? eg. showAdded, since it shows that the task is added.

jeanyang04 commented on src/main/java/oliver/Ui.java:
```diff
@@ -0,0 +1,165 @@
+     *
+     * @return string representation of the error message
+     */
+    public String showMissingArgsError() {
+        return "Missing arguments for this command.";
+    }
+
+    /**
+     * Shows the error message when invalid arguments are provided for a command.
+     *
+     * @return string representation of the error message
+     */
+    public String showInvalidArgsError() {
+        return "Invalid arguments provided for this command.";
+    }
+
+    /**
+     * Shows the error message when arguments provided for a command are out of range.
+     *
+     * @return string representation of the error message
+     */
+    public String showOutOfRangeError() {
+        return "No such task exists. Task number out of range.";
+    }
+}
```
Perhaps we could use a custom exception class to handle all of these, which may help simplify the logic of your Ui class.

jeanyang04 reviewed (commented):
Overall you did a great job! Some suggestions would be to incorporate a custom exception class.

SQ77 committed:
Add assertions to document important assumptions

The code previously did not check if the last character was a newline
before attempting to remove it, which could lead to unexpected behavior
or errors if the assumption was incorrect.

To prevent this, assertions are added to ensure the last character is a
newline before it is removed from the StringBuilder.

Let's,
* ensure better code safety and clarity by validating assumptions

SQ77 committed:
Refactor code to improve overall quality

The existing codebase contains sections that could benefit from
refactoring to enhance readability, reduce complexity, and improve
maintainability. By examining the code critically, we identify and
address these areas for improvement.

Let's,
* restructure complex methods for better readability
* eliminate redundant code and improve efficiency
* apply best practices for clearer and more concise logic

SQ77 committed:
Merge pull request #1 from SQ77/branch-A-Assertions

Add assertions to document important assumptions

SQ77 committed:
Merge branch 'master' into branch-A-CodeQuality

SQ77 committed:
Merge pull request #2 from SQ77/branch-A-CodeQuality

Refactor code to improve overall quality

SQ77 committed:
Remind users of upcoming tasks

SQ77 committed:
Merge branch 'branch-B-Reminders'

SQ77 committed:
Abstract out error printing

SQ77 committed:
Update Oliver's response for more personality

SQ77 committed:
Merge branch 'branch-A-Personality'

SQ77 committed:
Fix bug to display welcome correctly

SQ77 committed:
Improve the GUI of oliver

SQ77 committed:
Merge branch 'branch-A-BetterGui'

SQ77 committed:
Add draft version of user guide

SQ77 committed:
Update user guide with more details

SQ77 committed:
Change default README file

SQ77 committed:
Add more details to user guide
